proposed/http-message.md

```diff
@@ -332,9 +505,11 @@ interface IncomingRequestInterface extends MessageInterface
      * purposes of how duplicate query parameters are handled, and how nested
      * sets are handled.
      *
-     * @return array
+     * @param array $query Array of query string arguments, typically from
+     *     $_GET.
+     * @return void
      */
-    public function getQueryParams();
+    public function setQueryParams(array $query);
```

Comment:
We've had a few revisions. In the original `IncomingRequestInterface` implementation, it was mutable. Then we went with the "selective mutability", and it wasn't. Then I made an argument for essentially mutability of all input sources other than `$_SERVER` and `$_FILES`, with the assumption that you can rebuild the original state of the input sources from those and the content body.

Interestingly, `$_GET` is mutable. My assertion is that query parameters should reflect what PHP does here -- and thus also be mutable.

Altering `$_GET` does not alter either `$_SERVER['REQUEST_URI']` or `$_SERVER['QUERY_STRING']`, so my take is that changes to the query parameters in the request object should not update the request URI either. I'll document that.